server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java

```diff
@@ -172,6 +176,73 @@ public void collect(int doc, long bucket) throws IOException {
         directory.close();
     }
 
+    /**
+     * Check that knn queries rewrite parallelizes on the number of segments if there are enough
+     * threads available.
+     */
+    public void testConcurrentKnnRewrite() throws Exception {
+        final Directory directory = newDirectory();
+        try (
+            IndexWriter iw = new IndexWriter(
+                directory,
+                new IndexWriterConfig(new StandardAnalyzer()).setMergePolicy(NoMergePolicy.INSTANCE)
+            )
+        ) {
+            final int numDocs = randomIntBetween(100, 200);
+            for (int i = 0; i < numDocs; i++) {
+                Document document = new Document();
+                document.add(new StringField("field", "value", Field.Store.NO));
+                iw.addDocument(document);
+                if (rarely()) {
+                    iw.commit();
+                }
+            }
+        }
+
+        DirectoryReader directoryReader = DirectoryReader.open(directory);
+        int numSegments = directoryReader.getContext().leaves().size();
+        // make sure we have more threads than segments available to check later call to execute method
+        int nThreads = randomIntBetween(numSegments, numSegments + 5);
+
+        // use an executor that counts calls to its "execute" method
+        AtomicInteger executeCalls = new AtomicInteger(0);
+        ThreadPoolExecutor executor = new ThreadPoolExecutor(
+            nThreads,
+            nThreads,
+            0L,
+            TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<Runnable>()
+        ) {
+            @Override
+            public void execute(Runnable command) {
+                executeCalls.incrementAndGet();
+                super.execute(command);
+            }
+        };
+
+        ContextIndexSearcher searcher = new ContextIndexSearcher(
+            directoryReader,
+            IndexSearcher.getDefaultSimilarity(),
+            IndexSearcher.getDefaultQueryCache(),
+            IndexSearcher.getDefaultQueryCachingPolicy(),
+            1,
+            randomBoolean(),
+            executor
+        );
+        // check that we calculate one slice per segment
+        assertEquals(numSegments, searcher.slices(directoryReader.getContext().leaves()).length);
+
+        KnnFloatVectorQuery vectorQuery = new KnnFloatVectorQuery("float_vector", new float[] { 0, 0, 0 }, 10, null);
+        vectorQuery.rewrite(searcher);
```

Comment:
you may as well want to do a search here, and test how many tasks get created for that. Provide the rewritten query so that rewrite as part of search is a no-op.